### ISSUE TITLE
Dynamic monitor setup

### ISSUE THREAD

**Issue by jokogr** [hyprwm/Hyprland#1719]
I am using Hyprland on my laptop and I am used to have a dynamic monitor setup: When I am at my desk, I prefer to disable the laptop's monitor and use an external one instead. When I'm not sitting there, I just use the laptop's one obviously.

In Sway I have been using Kanshi and it has been working, but in Hyprland it appears to be crashing. My current workaround is to treat the situation manually:

Whenever I am plugging in a monitor, I disable the laptop's one via `hyprctl keyword monitor "eDP-1,disable"` and whenever I am about to unplug the former, I run `hyprctl keyword monitor "eDP-1, highrr, auto, 2"` so that the system is not left without a monitor after the unplugging.

Is there a way to support this workflow automatically? E.g., just before the last monitor gets unplugged, Hyprland re-scans for monitors and re-adds them if needed.

**Comment by vaxerski** [comment]
You can write a script that listens to [socket2](https://wiki.hyprland.org/IPC/) for the monitoradded/removed events and reacts with hyprctl calls.

**Comment by jokogr** [comment]
So I made a naive implementation where I am trying to bring up the laptop's monitor if the last one is removed: 

```
function handle {
  readarray -t inputArr < <(echo "${1//>>/$'\n'}")
  if [[ ${#inputArr[@]} != 2 ]]; then
    echo "Unrecognized format: $1"
    return
  fi
  event=${inputArr[0]}
  data=${inputArr[1]}
  if [ "$event" == "monitorremoved" ]; then
    if [[ $(hyprctl -j monitors | jq length) == 1 ]] && [ "$(hyprctl -j monitors | jq -r '.[0].name')" == "$data" ]; then
      hyprctl keyword monitor "eDP-1, highrr, auto, 2"
    fi
    return
  fi
}
```

My scenario: I have manually disabled the laptop's monitor (eDP-1), there is only DP-1 enabled and I am about to unplug it.

Unfortunately this leads to crash.

Here are some logs:

```
[LOG] Applying monitor rule for eDP-1
[WARN] No rule found for DP-3, trying to use the first.
[WARN] No rules configured. Using the default hardcoded one.
[LOG] Applying monitor rule for DP-3
[LOG] Setting preferred mode for DP-3
[LOG] Monitor DP-3 -> destroyed all render data
[LOG] LayerSurface 20c09a0 arranged: x: 3840 y: 0 w: 3440 h: 26 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor DP-3 layers arranged: reserved: 0.000000 26.000000 0.000000 0.000000
[LOG] Monitor DP-3 data dump: res 3440x1440@49.99Hz, scale 1.00, transform 0, pos 3840x0, 10b 0
[LOG] LayerSurface 20c09a0 arranged: x: 3840 y: 0 w: 3440 h: 26 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor DP-3 layers arranged: reserved: 0.000000 26.000000 0.000000 0.000000
[LOG] Fading out a non-existent LS??
[LOG] Framebuffer created, status 36053
[LOG] Framebuffer created, status 36053
[LOG] Framebuffer created, status 36053
[LOG] Framebuffer created, status 36053
[LOG] Allocated texture for BGTex
[LOG] Background created for monitor DP-3
[LOG] Shaders initialized successfully.
[LOG] Keybind triggered, calling dispatcher (64, , 50)
[LOG] Changed to workspace 2
[LOG] Set keyboard focus to surface 33cfd90, with window name: zsh
[LOG] Destroying workspace ID 1
[LOG] Framebuffer created, status 36053
[LOG] Callback 1634098 -> 1634090, Keyboard removed.
[LOG] Callback 1633f60 -> 1633f58, Keyboard removed.
[LOG] Callback 1633fc8 -> 1633fc0, Keyboard removed.
[LOG] Callback 1634030 -> 1634028, Keyboard removed.
[LOG] Destroyed keyboard 1633f30
[LOG] Callback 166db88 -> 166db80, Keyboard removed.
[LOG] Callback 166da50 -> 166da48, Keyboard removed.
[LOG] Callback 166dab8 -> 166dab0, Keyboard removed.
[LOG] Callback 166db20 -> 166db18, Keyboard removed.
[LOG] Destroyed keyboard 166da20
[LOG] Callback 17b2198 -> 17b2190, Keyboard removed.
[LOG] Callback 17b2060 -> 17b2058, Keyboard removed.
[LOG] Callback 17b20c8 -> 17b20c0, Keyboard removed.
[LOG] Callback 17b2130 -> 17b2128, Keyboard removed.
[LOG] Destroyed keyboard 17b2030
[LOG] Callback 16a72d8 -> 16a72d0, Keyboard removed.
[LOG] Callback 16a71a0 -> 16a7198, Keyboard removed.
[LOG] Callback 16a7208 -> 16a7200, Keyboard removed.
[LOG] Callback 16a7270 -> 16a7268, Keyboard removed.
[LOG] Destroyed keyboard 16a7170
[LOG] Callback 15afa08 -> 15afa00, Mouse removed.
[LOG] Destroy called for monitor DP-3
[LOG] onDisconnect called for DP-3
[LOG] Callback 1873d58 -> 1873d50,  removed.
[LOG] LayerSurface 20bcfd0 unmapped
[LOG] Framebuffer created, status 36053
[LOG] LayerSurface 20bcfd0 destroyed
[LOG] Callback 1c76eb8 -> 1c76eb0, layerSurface removed.
[LOG] Callback 1c76d80 -> 1c76d78, layerSurface removed.
[LOG] Callback 1c76de8 -> 1c76de0, layerSurface removed.
[LOG] Callback 1c76e50 -> 1c76e48, layerSurface removed.
[LOG] Callback 1c76f20 -> 1c76f18, layerSurface removed.
[LOG] Monitor DP-3 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Removed monitor DP-3!
[WARN] Unplugged last monitor, entering an unsafe state. Good luck my friend.
[LOG] Callback 1873e28 -> 1873e20,  removed.
[LOG] Callback 1873dc0 -> 1873db8,  removed.
[LOG] Hyprctl: keyword monitor : eDP-1, highrr, auto, 2
[LOG] Applying monitor rule for eDP-1
[LOG] Callback 1852c10 -> 1852c08,  removed.
[LOG] Callback 1852c78 -> 1852c70,  removed.
[LOG] Registered signal for owner 1852a30: 1838760 -> 1852b80 (owner: )
[LOG] Registered signal for owner 1852a30: 18387f0 -> 1852be8 (owner: )
[LOG] Registered signal for owner 1852a30: 18387e0 -> 1852c50 (owner: )
[LOG] Added new monitor with name eDP-1 at -1,-1 with size 3840x2400, pointer 1838610
[LOG] New monitor: WORKSPACEID 1, exists: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Monitor eDP-1: Applying highest mode 3840x2400@60.000000.
[LOG] Monitor eDP-1 -> destroyed all render data
```

Let me know if I should file this as a bug

**Comment by vaxerski** [comment]
can you post the crash report?

**Comment by jokogr** [comment]
I would like to omit some system and release information if that's OK:

```
--------------------------------------------
   Hyprland Crash Report
--------------------------------------------
Who tried dividing by zero?!

Hyprland received signal 11 (Segmentation fault)

(omitted system info)

GPU:
	00:02.0 VGA compatible controller [0300]: Intel Corporation TigerLake-LP GT2 [Iris Xe Graphics] [8086:9a49] (rev 01) (prog-if 00 [VGA controller])


(omitted os-release)
	


Backtrace:
	#0 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x4d025d]
		_ZN13CrashReporter18createAndSaveCrashEi
		??:?
	#1 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x459e0f]
		_Z25handleUnrecoverableSignali
		??:?
	#2 | /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6(+0x3dbf0) [0x7f9d8a03dbf0]
		??
		??:0
	#3 | /nix/store/sg275pvghxp61hxyirxhz0jl3617f5hh-wayland-1.21.0/lib/libwayland-server.so.0(wl_list_insert+0x5) [0x7f9d8ade6435]
		??
		??:0
	#4 | /nix/store/7ids47lamf5pk3831qb82fpkfm42kf35-wlroots-hyprland-hidpi-2023-03-02_5ae17de/lib/libwlroots.so.12(wlr_output_configuration_head_v1_create+0x4f) [0x7f9d8ad4fcbf]
		??
		??:0
	#5 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x52d806]
		_ZN6Events15listener_changeEP11wl_listenerPv
		??:?
	#6 | /nix/store/sg275pvghxp61hxyirxhz0jl3617f5hh-wayland-1.21.0/lib/libwayland-server.so.0(wl_signal_emit_mutable+0x7c) [0x7f9d8ade18fc]
		??
		??:0
	#7 | /nix/store/7ids47lamf5pk3831qb82fpkfm42kf35-wlroots-hyprland-hidpi-2023-03-02_5ae17de/lib/libwlroots.so.12(+0x89b65) [0x7f9d8ad4db65]
		??
		??:0
	#8 | /nix/store/7ids47lamf5pk3831qb82fpkfm42kf35-wlroots-hyprland-hidpi-2023-03-02_5ae17de/lib/libwlroots.so.12(+0x89dd8) [0x7f9d8ad4ddd8]
		??
		??:0
	#9 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x4fb0c3]
		_ZN13CHyprRenderer16applyMonitorRuleEP8CMonitorP12SMonitorRuleb
		??:?
	#10 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x501d83]
		_ZN14CConfigManager10ensureDPMSEv
		??:?
	#11 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x4e213d]
		_Z13hyprCtlFDTickijPv
		??:?
	#12 | /nix/store/sg275pvghxp61hxyirxhz0jl3617f5hh-wayland-1.21.0/lib/libwayland-server.so.0(wl_event_loop_dispatch+0xa2) [0x7f9d8ade37c2]
		??
		??:0
	#13 | /nix/store/sg275pvghxp61hxyirxhz0jl3617f5hh-wayland-1.21.0/lib/libwayland-server.so.0(wl_display_run+0x25) [0x7f9d8ade1335]
		??
		??:0
	#14 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x42e5fc]
		main
		??:?
	#15 | /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6(+0x2924e) [0x7f9d8a02924e]
		??
		??:0
	#16 | /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6(__libc_start_main+0x89) [0x7f9d8a029309]
		??
		??:0
	#17 | /nix/store/w12wsn6p90hdkn7vv0bscx4r08s0dp2x-hyprland-0.22.0+date=2023-03-06_4f647a8/bin/Hyprland() [0x443795]
		_start
		??:?


Log tail:
[LOG] Shaders initialized successfully.
[LOG] Callback 2f89f98 -> 2f89f90, Keyboard removed.
[LOG] Callback 2f89e60 -> 2f89e58, Keyboard removed.
[LOG] Callback 2f89ec8 -> 2f89ec0, Keyboard removed.
[LOG] Callback 2f89f30 -> 2f89f28, Keyboard removed.
[LOG] Destroyed keyboard 2f89e30
[LOG] Callback 2fb04b8 -> 2fb04b0, Keyboard removed.
[LOG] Callback 2fb0380 -> 2fb0378, Keyboard removed.
[LOG] Callback 2fb03e8 -> 2fb03e0, Keyboard removed.
[LOG] Callback 2fb0450 -> 2fb0448, Keyboard removed.
[LOG] Destroyed keyboard 2fb0350
[LOG] Callback 311b908 -> 311b900, Keyboard removed.
[LOG] Callback 311b7d0 -> 311b7c8, Keyboard removed.
[LOG] Callback 311b838 -> 311b830, Keyboard removed.
[LOG] Callback 311b8a0 -> 311b898, Keyboard removed.
[LOG] Destroyed keyboard 311b7a0
[LOG] Callback 2ffbf38 -> 2ffbf30, Keyboard removed.
[LOG] Callback 2ffbe00 -> 2ffbdf8, Keyboard removed.
[LOG] Callback 2ffbe68 -> 2ffbe60, Keyboard removed.
[LOG] Callback 2ffbed0 -> 2ffbec8, Keyboard removed.
[LOG] Destroyed keyboard 2ffbdd0
[LOG] Callback 3028e48 -> 3028e40, Mouse removed.
[LOG] Destroy called for monitor DP-3
[LOG] onDisconnect called for DP-3
[LOG] Callback 31c83c8 -> 31c83c0,  removed.
[LOG] LayerSurface 3a09a80 unmapped
[LOG] Framebuffer created, status 36053
[LOG] LayerSurface 3a09a80 destroyed
[LOG] Callback 35c86f8 -> 35c86f0, layerSurface removed.
[LOG] Callback 35c85c0 -> 35c85b8, layerSurface removed.
[LOG] Callback 35c8628 -> 35c8620, layerSurface removed.
[LOG] Callback 35c8690 -> 35c8688, layerSurface removed.
[LOG] Callback 35c8760 -> 35c8758, layerSurface removed.
[LOG] Monitor DP-3 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Removed monitor DP-3!
[WARN] Unplugged last monitor, entering an unsafe state. Good luck my friend.
[LOG] Callback 31c8498 -> 31c8490,  removed.
[LOG] Callback 31c8430 -> 31c8428,  removed.
[LOG] Hyprctl: keyword monitor : eDP-1, highrr, auto, 2
[LOG] Applying monitor rule for eDP-1
[LOG] Callback 31a7590 -> 31a7588,  removed.
[LOG] Callback 31a75f8 -> 31a75f0,  removed.
[LOG] Registered signal for owner 31a73b0: 318d1e0 -> 31a7500 (owner: )
[LOG] Registered signal for owner 31a73b0: 318d270 -> 31a7568 (owner: )
[LOG] Registered signal for owner 31a73b0: 318d260 -> 31a75d0 (owner: )
[LOG] Added new monitor with name eDP-1 at -1,-1 with size 3840x2400, pointer 318d090
[LOG] New monitor: WORKSPACEID 2, exists: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Monitor eDP-1: Applying highest mode 3840x2400@60.000000.
[LOG] Monitor eDP-1 -> destroyed all render data
```

**Comment by vaxerski** [comment]
looks like #804 then

**Comment by DrummyFloyd** [comment]
any update regarding this behaviour ? because i have the same issue

**Comment by Ramblurr** [comment]
@DrummyFloyd this issue is closed because it is a dupe of #804. To be notified of updates to that issue, browse to it then click the bell icon to be notified of changes.
